lib/Sema/CSDiag.cpp

```diff
@@ -6508,7 +6510,104 @@ bool FailureDiagnosis::visitClosureExpr(ClosureExpr *CE) {
+                         argTupleTy);
+                return true;
+              }
+            }
+
+            auto diag = diagnose(params->getStartLoc(),
+                                 diag::closure_tuple_parameter_destructuring,
+                                 argTupleTy);
+            Type actualArgType;
+            if (auto *actualFnType = CE->getType()->getAs<AnyFunctionType>())
+              actualArgType = actualFnType->getInput();
+
+            auto &sourceMgr = CS->getASTContext().SourceMgr;
+            auto paramRange = CharSourceRange(
+                sourceMgr, params->getStartLoc(),
+                Lexer::getLocForEndOfToken(sourceMgr, params->getEndLoc()));
+
+            SmallString<64> fixIt;
+            llvm::raw_svector_ostream OS(fixIt);
+
+            // If this is multi-line closure we'd have to insert new lines
+            // in the suggested 'let' to keep the structure of the code in
+            // tact, otherwise just use ';' to keep everything on the same
+            // line.
+            bool isMultiLineClosure = false;
+            if (auto *body = CE->getBody()) {
+              auto inLine = sourceMgr.getLineNumber(CE->getInLoc());
+              auto bodyLine = sourceMgr.getLineNumber(body->getStartLoc());
+              isMultiLineClosure = bodyLine > inLine;
+            }
+
+            std::string parameter;
+            auto paramText = sourceMgr.extractText(paramRange);
+            // If parameter is not enclosed in parenthesis,
+            // let's add them since we are forming a tuple.
+            if (!paramText.startswith("(")) {
+              parameter = "(" + paramText.str() + ")";
+            } else {
+              parameter = paramText;
+            }
+
+            // If at least one of the parameters has explicitly
+            // declared type let's turn fix-it into N 'let' expressions
+            // to be able to specify it.
+            bool explicitTypes = false;
+            for (auto *param : params->getArray()) {
+              if (!param->getType()->is<TypeVariableType>()) {
+                explicitTypes = true;
+                break;
+              }
+            }
+
+            if (isMultiLineClosure || explicitTypes)
+              OS << '\n';
+
+            // Let's form 'let <name> : <type> = arg.N' expression.
+            if (explicitTypes) {
+              for (unsigned i = 0, e = params->size(); i != e; ++i) {
+                auto *param = params->get(i);
+                if (param->getType()->is<TypeVariableType>()) {
+                  OS << "let " << param->getNameStr();
+                } else {
+                  param->print(OS);
+                }
+
+                OS << " = arg." << i << '\n';
+              }
+            } else {
+              OS << "let " << parameter << " = arg"
+                 << (isMultiLineClosure ? "\n" : "; ");
+            }
+
+            diag.fixItReplace(params->getSourceRange(), "arg")
```

Comment:
If we're starting with a single-expression closure, this probably should include the type (`(arg: (Int, Double))`). It also should include the return type if hasn't already been specified and isn't Void.

If we're starting with a multi-statement closure, the return type is fine, but the arguments should still have a type if any of the parameters-as-written had a type.